File: lib/csharp/microsoft.bot.builder.skills/Microsoft.Bot.Builder.Skills/Auth/MSJwtAuthenticationProvider.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.IdentityModel.Tokens.Jwt;
using System.Linq;
using System.Security.Claims;
using System.Threading;
using Microsoft.IdentityModel.Protocols;
using Microsoft.IdentityModel.Protocols.OpenIdConnect;
using Microsoft.IdentityModel.Tokens;

namespace Microsoft.Bot.Builder.Skills.Auth
{
    public class MSJwtAuthenticationProvider : IAuthenticationProvider
    {
        private static OpenIdConnectConfiguration _openIdConfig;
        private readonly string _microsoftAppId;
        private readonly string _openIdMetadataUrl = "https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration";

        public MSJwtAuthenticationProvider(string microsoftAppId, string openIdMetadataUrl = null)
        {
            _microsoftAppId = !string.IsNullOrWhiteSpace(microsoftAppId) ? microsoftAppId : throw new ArgumentNullException(nameof(microsoftAppId));
            if (!string.IsNullOrWhiteSpace(openIdMetadataUrl))
            {
                _openIdMetadataUrl = openIdMetadataUrl;
            }

            var configurationManager = new ConfigurationManager<OpenIdConnectConfiguration>(_openIdMetadataUrl, new OpenIdConnectConfigurationRetriever());
            _openIdConfig = configurationManager.GetConfigurationAsync(CancellationToken.None).GetAwaiter().GetResult();
        }

        public ClaimsIdentity Authenticate(string authHeader)
        {
            if (authHeader == null)
            {
                throw new ArgumentNullException(nameof(authHeader));
            }

            var validationParameters = new TokenValidationParameters
            {
                ValidateIssuer = false, // do not validate issuer
                ValidAudiences = new[] { _microsoftAppId },
                IssuerSigningKeys = _openIdConfig.SigningKeys,
            };

            var handler = new JwtSecurityTokenHandler();
            var user = handler.ValidateToken(authHeader.Replace("Bearer ", string.Empty), validationParameters, out _);

            return user.Identities.FirstOrDefault();
        }
    }
}

File: sdk/typescript/libraries/botbuilder-skills/src/auth/msJWTAuthenticationProvider.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

// Extends verify definitions to be compatible with callback handler to resolve signingKey
declare module 'jsonwebtoken' {
    export type signingKeyResolver = (headers: jwks.Headers, cb: (err: Error, signingKey: string) => void) => void;

    export function verify(
        token: string,
        secretOrPublicKey: signingKeyResolver,
        callback?: VerifyCallback
    ): void;
}

import { ClaimsIdentity } from 'botframework-connector';
import * as jwks from 'jwks-rsa';
import { IAuthenticationProvider } from './authenticationProvider';

export class MsJWTAuthenticationProvider implements IAuthenticationProvider {
    // private static openIdConfig: OpenIdConnectConfiguration;
    private readonly microsoftAppId: string;
    private readonly openIdMetadataUrl: string = 'https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration';

    public constructor (microsoftAppId: string, openIdMetadataUrl: string = '') {
        if (microsoftAppId === undefined || microsoftAppId.trim().length === 0) { throw new Error('MicrosoftAppId is undefined'); }
        this.microsoftAppId = microsoftAppId;

        if (openIdMetadataUrl !== undefined && openIdMetadataUrl.trim().length > 0) {
            this.openIdMetadataUrl = openIdMetadataUrl;
        }
        /* PENDING: ConfigurationManager and IdentityModel is not present in JS/TS
            var configurationManager = new ConfigurationManager<OpenIdConnectConfiguration>(
                _openIdMetadataUrl,
                new OpenIdConnectConfigurationRetriever()
                );
            openIdConfig = configurationManager.GetConfigurationAsync(CancellationToken.None).GetAwaiter().GetResult();
        */
    }

    public authenticate(authHeader: string): ClaimsIdentity {
        /* PENDING: IdentityModel is not present in JS/TS
            try
            {
                var validationParameters =
                    new TokenValidationParameters
                    {
                        ValidateIssuer = false, // do not validate issuer
                        ValidAudiences = new[] { _microsoftAppId },
                        IssuerSigningKeys = openIdConfig.SigningKeys,
                    };

                var handler = new JwtSecurityTokenHandler();
                var user = handler.ValidateToken(authHeader.Replace("Bearer ", string.Empty), validationParameters, out var validatedToken);

                return user.Identities.OfType<ClaimsIdentity>().FirstOrDefault();
            }
            catch
            {
                return null;
            }
        */

        return new ClaimsIdentity([], false);
    }
}
